cirq-core/cirq/transformers/transformer_api.py

```diff
@@ -0,0 +1,327 @@
+          would end (i.e. receive a `register_final` call) once all nested transformers (i.e. all
+          `register_initial` calls received while the top level transformer was active) have
+          finished (i.e. corresponding `register_final` calls have also been received).
+        - This behavior can be simulated by maintaining a stack of currently active stages and
+          adding data from `log` calls to the stage at the top of the stack.
+
+    The `LogLevel`s can be used to control the input processing and output resolution of the logs.
+    """
+
+    def __init__(self):
+        """Initializes TransformerStatsLogger."""
+        self._curr_id: int = 0
+        self._logs: List[LoggerNode] = []
+        self._stack: List[int] = []
+
+    def register_initial(self, circuit: 'cirq.AbstractCircuit', transformer_name: str) -> None:
+        """Register the beginning of a new transformer stage.
+
+        Args:
+            circuit: Input circuit to the new transformer stage.
+            transformer_name: Name of the new transformer stage.
+        """
+        if self._stack:
+            self._logs[self._stack[-1]].nested_loggers += (self._curr_id,)
+        self._logs.append(LoggerNode(self._curr_id, transformer_name, circuit, circuit))
+        self._stack.append(self._curr_id)
+        self._curr_id += 1
+
+    def log(self, *args: str, level: LogLevel = LogLevel.INFO) -> None:
+        """Log additional metadata corresponding to the currently active transformer stage.
+
+        Args:
+            *args: The additional metadata to log.
+            level: Logging level to control the amount of metadata that gets put into the context.
+
+        Raises:
+            ValueError: If there's no active transformer on the stack.
+        """
+        if len(self._stack) == 0:
+            raise ValueError('No active transformer found.')
+        self._logs[self._stack[-1]].logs += ((level, args),)
+
+    def register_final(self, circuit: CIRCUIT_TYPE, transformer_name: str) -> None:
+        """Register the end of the currently active transformer stage.
+
+        Args:
+            circuit: Final transformed output circuit from the transformer stage.
+            transformer_name: Name of the (currently active) transformer stage which ends.
+
+        Raises:
+            ValueError: If `transformer_name` is different from currently active transformer name.
+        """
+        tid = self._stack.pop()
+        if self._logs[tid].transformer_name != transformer_name:
+            raise ValueError(
+                f"Expected `register_final` call for currently active transformer "
+                f"{self._logs[tid].transformer_name}."
+            )
+        self._logs[tid].final_circuit = circuit
+
+    def show(self, level: LogLevel = LogLevel.INFO) -> None:
+        """Show the stored logs >= level in the desired format."""
+
+        def print_log(log: LoggerNode, pad=''):
```

Comment:
Nit: Can we move this out to a protected class level function with a leading underscore instead of a nested function ?

Reply:
I chose this to be a nested method because derived classes will be overriding the `show` method and then the inherited `_print` will become moot.